### PR TITLE
TST: Use pytest instead of nose

### PR DESCRIPTION
**Pull request by QuLogic**
Since nose is [sort of dead](https://nose.readthedocs.org/en/latest/#note-to-users) and I just saw [nose2pytest](https://github.com/schollii/nose2pytest) and [unittest2pytest](https://github.com/pytest-dev/unittest2pytest), I thought I'd try them out just to see how they'd work.

I've added the straight-forward changes to plain `assert` statements. What's remaining are ~~the assertions for raised exceptions,~~ plus maybe using py.test fixtures. ~~At the moment, there are no hard dependencies on py.test, and the exception assertions still use nose.~~

### CODE REVIEW

**Comment by QuLogic** [comment]
I can't figure out why pytest is trying to import the shared libraries on Travis, but not locally.

**Comment by QuLogic** [comment]
So, using `pip` instead of `setup.py` fixed up Travis.

File: .travis.yml [hunk 1]
```diff
@@ -42,8 +42,9 @@ install:
   # Customise the testing environment
   # ---------------------------------
   - conda config --add channels scitools
-  - PACKAGES="$PACKAGES cython pillow nose pep8 proj4 pyshp shapely six requests pyepsg owslib"
+  - PACKAGES="$PACKAGES cython pillow nose pytest pep8 proj4 pyshp shapely six requests pyepsg owslib"
```
Comment (pelson) [inline .travis.yml]: can we remove nose from the list?

**Comment by pelson** [comment]
It's a brave new world without my nose familiarity, but I'm certainly :+1: on moving towards another fairly ubiquitous testing tool.

File: .travis.yml (hunk 1 above)
Comment (QuLogic) [inline .travis.yml]: No, it's still need for matplotlib's [image comparison decorator](https://travis-ci.org/SciTools/cartopy/jobs/121297187#L752-L757).

**Comment by ajdawson** [comment]
I'd like to move forward with this one, are the conflicts easy to fix?

**Comment by QuLogic** [comment]
Seems to be mostly imports with one or two smaller test function changes, but I haven't tried to rebase recently.

**Comment by QuLogic** [comment]
Probably should wait until after #773.

**Comment by ajdawson** [comment]
OK, #773 is in, let us know how you get on.

**Comment by ajdawson** [comment]
If all goes well, can you add pytest to the testing dependencies section of INSTALL too.

**Comment by QuLogic** [comment]
Travis had a few connection issues, but this should be good now. I also marked the MapQuest tests as `xfail`, so they don't fail the whole build. They could also be marked as `skip` instead if running them it too costly.

**Comment by ajdawson** [comment]
This looks OK. I note that the total number of tests reported running is smaller than before the PR, is that due to the way tests are counted, or actual removal/refactoring of tests (or is it an error)?

File: requirements/tests.txt [hunk 2]
```diff
@@ -1,3 +1,3 @@
 mock>=1.0.1
-nose>=1.2.1
```
Comment (ajdawson) [inline requirements/tests.txt]: Can you clarify if we actually need nose still?

File: INSTALL [hunk 3]
```diff
@@ -114,7 +114,7 @@ These packages are required for the full Cartopy test suite to run.
     Python mocking and patching package for testing. Note that this package
     is only required to support the Cartopy unit tests.
 
-**nose** 1.2.1 or later (https://nose.readthedocs.org/en/latest/)
```
Comment (ajdawson) [inline INSTALL]: Is nose still required due to the mpl image testing?

File: requirements/tests.txt (hunk 2 above)
Comment (QuLogic) [inline requirements/tests.txt]: We don't directly require it.

File: INSTALL (hunk 3 above)
Comment (QuLogic) [inline INSTALL]: Yes, depending on version; 2.1 should not require it when it's released.

Comment (ajdawson) [inline INSTALL]: OK cheers

**Comment by QuLogic** [comment]
> I note that the total number of tests reported running is smaller than before the PR,

There seems to be a difference between the old dependencies build and the latest everything build. The old one is missing 6 tests:
```
cartopy/tests/mpl/test_axes.py::test_geoaxes_subplot
cartopy/tests/mpl/test_contour.py::test_contour_plot_bounds
cartopy/tests/mpl/test_quiver.py::TestQuiverShapes::test_quiver_transform_xy_1d_uv_2d
cartopy/tests/mpl/test_quiver.py::TestQuiverShapes::test_quiver_transform_xyuv_1d
cartopy/tests/mpl/test_set_extent.py::test_domain_extents
cartopy/tests/mpl/test_set_extent.py::test_extents
```
I'm not entirely sure what's special about these tests so far.

**Comment by ajdawson** [comment]
I think that was sort-of true before this PR though? I was referring to the following:

Before: 323/325 tests
After: 279/285 tests

**Comment by QuLogic** [comment]
For the six differences, it's because of the `cleanup` decorator, which I think must be broken in old Matplotlib and just cannot be fixed anyway.

For the remaining differences you mentioned, the last commit should fix it. Some tests classes were incorrectly named, preventing collection.

**Comment by ajdawson** [comment]
OK I'm happy with that. Thanks @QuLogic, good work!

**Comment by pelson** [comment]
Wow. Just going through my backlog - what a great change. Thanks @QuLogic & thanks @ajdawson for moving it forward!

**Pull request merged**